src/commands/text.ts

```diff
@@ -0,0 +1,470 @@
+
+    if (anticursor[dir] === this) {
+      var newPc = new TextPiece(ch).createDir(dir, cursor);
+      anticursor[dir] = newPc;
+      cursor.insDirOf(dir, newPc);
+    }
+    else {
+      var from = this[-dir as Direction];
+      if (from instanceof TextPiece) from.insTextAtDirEnd(ch, dir);
+      else {
+        var newPc = new TextPiece(ch).createDir(-dir as Direction, cursor);
+        var selection = cursor.selection;
+        if (selection) {
+          newPc.jQ.insDirOf(-dir as Direction, selection.jQ);
+        }
+      }
+
+      if (this.textStr.length === 1 && anticursor[-dir as Direction] === this) {
+        anticursor[-dir as Direction] = this[-dir as Direction]; // `this` will be removed in deleteTowards
+      }
+    }
+
+    return this.deleteTowards(dir, cursor);
+  };
+};
+
+LatexCmds.text =
+LatexCmds.textnormal =
+LatexCmds.textrm =
+LatexCmds.textup =
+LatexCmds.textmd = TextBlock;
+
+function makeTextBlock(latex:string, ariaLabel:string, tagName:string, attrs:string) {
+  return class extends TextBlock {
+    ctrlSeq = latex;
+    mathspeakTemplate = ['Start'+ariaLabel, 'End'+ariaLabel];
+    ariaLabel = ariaLabel;
+
+    html () {
+      var cmdId = 'mathquill-command-id=' + this.id;
+      return '<'+tagName+' '+attrs+' '+cmdId+'>'+this.textContents()+'</'+tagName+'>';
+    }
+  };
+}
+
+LatexCmds.em = LatexCmds.italic = LatexCmds.italics =
+LatexCmds.emph = LatexCmds.textit = LatexCmds.textsl =
+  makeTextBlock('\\textit', 'Italic', 'i', 'class="mq-text-mode"');
+LatexCmds.strong = LatexCmds.bold = LatexCmds.textbf =
+  makeTextBlock('\\textbf', 'Bold', 'b', 'class="mq-text-mode"');
+LatexCmds.sf = LatexCmds.textsf =
+  makeTextBlock('\\textsf', 'Sans serif font', 'span', 'class="mq-sans-serif mq-text-mode"');
+LatexCmds.tt = LatexCmds.texttt =
+  makeTextBlock('\\texttt', 'Mono space font', 'span', 'class="mq-monospace mq-text-mode"');
+LatexCmds.textsc =
+  makeTextBlock('\\textsc', 'Variable font', 'span', 'style="font-variant:small-caps" class="mq-text-mode"');
+LatexCmds.uppercase =
+  makeTextBlock('\\uppercase', 'Uppercase', 'span', 'style="text-transform:uppercase" class="mq-text-mode"');
+LatexCmds.lowercase =
+  makeTextBlock('\\lowercase', 'Lowercase', 'span', 'style="text-transform:lowercase" class="mq-text-mode"');
+
+
+class RootMathCommand extends MathCommand {
+  cursor:Cursor;
+  constructor (cursor:Cursor) {
+    super('$');
+    this.cursor = cursor;
+  };
+  htmlTemplate = '<span class="mq-math-mode">&0</span>';
+  createBlocks () {
+    super.createBlocks()
+    const endsL = this.ends[L] as RootMathCommand; // TODO - how do we know this is a RootMathCommand?
+    endsL.cursor = this.cursor;
+    endsL.write = function(cursor:Cursor, ch:string) {
+      if (ch !== '$')
+        MathBlock.prototype.write.call(this, cursor, ch);
+      else if (this.isEmpty()) {
+        cursor.insRightOf(this.parent);
+        this.parent.deleteTowards(undefined!, cursor);
```

Comment:
What does undefined! mean?

Reply:
As noted above, I just learned what this is for in typescript.

https://stackoverflow.com/questions/42273853/in-typescript-what-is-the-exclamation-mark-bang-operator-when-dereferenci

Reply:
As I just mentioned, this seems like bad practice.  If you're going to pass `undefined` into the method, then adjust the parameters for the method to take `undefined` and make sure that method properly checks for it.